File: the-7-sided-die/scala/die.scala
import scala.util.Random
import scala.collection.mutable.Map

object Die {

    val r = new Random()

    def main (args: Array[String]): Unit = {
        Benchmark.run(the6SidedDie.roll)
        Benchmark.run(the7SidedDie.rollWithPartition)
        Benchmark.run(the7SidedDie.rollWithFormula)
    }

    object the6SidedDie {

        def roll(): Int = r.nextInt(6)+1
    }

    object the7SidedDie {

        def roll(): Int = r.nextInt(7)+1

        def rollWithPartition(): Int =  (the6SidedDie.roll, the6SidedDie.roll) 
        match {
            case (6,6) => rollWithPartition()
            case (_,6) => 7
            case (n,_) => n
        }

        def rollWithFormula(): Int = {

            val fstDraw = the6SidedDie.roll
            val sndDraw = the6SidedDie.roll

            if (fstDraw == 6 && sndDraw == 6) 
                rollWithFormula
            else
                ((fstDraw-1)*6 + sndDraw) % 7 + 1
        }
    }

    object Benchmark {

        def run(f: () => Int): Unit = {

            val res : Map[Int,Int] = Map()
            val t0 = System.nanoTime
            for (i <- 1 to 10000000){
                val draw = f()
                res(draw) = (res getOrElse (draw,0)) + 1
            }
            val t1 = System.nanoTime

            for ((k,v) <- res) printf("key: %s, value: %s\n", k, v) 
            println("Roll took %f seconds.".format((t1-t0)/10E9))
        }
    }
}

File: the-7-sided-die/python/die.py
#!/usr/bin/python3

import random as rd
import time as t

class the_6_sided_die:

    def roll():
        return rd.randint(1,6)

class the_7_sided_die:

    def roll():
        return rd.randint(1,7)

    def roll_with_partition():
        i,j=the_6_sided_die.roll(), the_6_sided_die.roll()
        if (j == 6):
            if (i == 6):
                return the_7_sided_die.roll_with_partition()
            else:
                return 7
        else:
            return i

    def roll_with_formula():
        i,j=the_6_sided_die.roll(), the_6_sided_die.roll()
        if (i == 6 & j == 6):
            return the_7_sided_die.roll_with_formula()
        else:
            return (((i-1)*6+j)%7+1)

class benchmark:

    def run(f):
        r={}
        start = t.time()
        for i in range(int(1E6)):
            tmp=f()
            r[tmp]=r.get(tmp,0)+1
        end = t.time()

        print("Rolling with",f.__name__,"took",round(end-start,2), "seconds. Distribution is :", sep=' ')
        for elt in r.items():
            print(elt)

benchmark.run(the_7_sided_die.roll)
benchmark.run(the_7_sided_die.roll_with_partition)
benchmark.run(the_7_sided_die.roll_with_formula)
